src/RevCPU.cc

```diff
@@ -770,128 +729,122 @@ void RevCPU::CheckBlockedThreads(){
-        
-        if( AssignedThreads.at(ProcID).empty() ){
-          Enabled[ProcID] = false;
-        }
+      output.verbose(CALL_INFO, 8, 0, "Thread %" PRIu32 "on Core %" PRIu32 " is BLOCKED\n", ThreadID, ProcID);
 
-      }
+      // Set its state to BLOCKED
+      Thread->SetState(ThreadState::BLOCKED);
+
+      // Add it to BlockedThreads
+      BlockedThreads.emplace_back(std::move(Thread));
       break;
-    case ThreadState::START: // Should never happen
-      output.verbose(CALL_INFO, 99, 1, "A new thread with ID = %" PRIu32 " was found on Core %" PRIu32, Thread->GetThreadID(), ProcID);
+    case ThreadState::START:
+      // A new thread was created
+      output.verbose(CALL_INFO, 99, 1, "A new thread with ID = %" PRIu32 " was found on Core %" PRIu32, Thread->GetID(), ProcID);
 
       // Mark it ready for execution
       Thread->SetState(ThreadState::READY);
 
-      // Add it to the Thread map
-      Threads.emplace(Thread->GetThreadID(), Thread);
+      // Add it to the ReadyThreads so it is scheduled
+      ReadyThreads.emplace_back(std::move(Thread));
 
-      // Add it to the thread queue to be scheduled
-      ThreadQueue.emplace_back(Thread->GetThreadID());
       break;
 
     case ThreadState::RUNNING:
-      output.verbose(CALL_INFO, 11, 0, "Thread %" PRIu32 " on Core %" PRIu32 " is RUNNING\n", Thread->GetThreadID(), ProcID);
+      output.verbose(CALL_INFO, 11, 0, "Thread %" PRIu32 " on Core %" PRIu32 " is RUNNING\n", ThreadID, ProcID);
       break;
 
     case ThreadState::READY:
       // If this happens we are not setting state when assigning thread somewhere
       output.fatal(CALL_INFO, 99, "Error: Thread %" PRIu32 " on Core %" PRIu32 " is assigned but is in READY state... This is a bug\n",
-                    Thread->GetThreadID(), ProcID);
+                    ThreadID, ProcID);
       break;
     default: // Should DEFINITELY never happen
-      output.fatal(CALL_INFO, 99, "Error: Thread %" PRIu32 " on Core %" PRIu32 " is in an unknown state... This is a bug\n",
-                    Thread->GetThreadID(), ProcID);
+      output.fatal(CALL_INFO, 99, "Error: Thread %" PRIu32 " on Core %" PRIu32 " is in an unknown state... This is a bug.\n%s\n",
+                    ThreadID, ProcID, Thread->to_string().c_str());
       break;
     }
-    // Pop the thread that changed state
-    Procs[ProcID]->GetThreadsThatChangedState().pop();
+    // If the Proc does not have any threads assigned to it, there is no work to do, it is disabled
+    if( Procs[ProcID]->HasNoBusyHarts() ){
+      Enabled[ProcID] = false;
+    }
+    // Remove the pointer to the thread as it was handled
+    ThreadsThatChangedState.pop();
   }
   return;
 }
 
+void RevCPU::InitMainThread(uint32_t MainThreadID, const uint64_t StartAddr){
+  // @Lee: Is there a better way to get the feature info?
+  std::unique_ptr<RevRegFile> MainThreadRegState = std::make_unique<RevRegFile>(Procs[0]->GetRevFeature());
```

Comment:
`Procs[0]` seems to be arbitrary. Is there a way to figure out which `Procs` to use here?

Reply:
It is semi-arbitrary. I used 0 because we ALWAYS know that there will be at least one Proc (Core ;)). Further, we also know that the main thread will always end up on the first core. 

I still don't like it... but currently the RevFeature is created inside of RevProc so that's the only way I can think to gather the info.

Reply:
Right now we assume a homogeneous CPU, so this is OK.  To get to a heterogeneous CPU (like a big.LITTLE) we'll need to do some rearchitecting and the CPU will need to have some description of what features (RV and other) are in which Procs. I don't think we need to solve that here, or hold up your PR for that reason